### PR TITLE
[Home] Conditionally show suggested rail.

### LINKED ISSUE #368: [Home] Determine what artist modules should be rendered.
* These artist modules should _always_ be shown, regardless of authentication status:
  - trending
  - iconic
* The ‘recommended for you’ artist rail should only be shown if:
  - a user is signed in
  - and they have at least 1 artist follow

### PR DESCRIPTION
**Pull request by alloy**
Fixes #368.

Still need to add actual popular artists, but need to finish some Gravity work for that first.

### CODE REVIEW

File: schema/home/home_page_artist_module.js [hunk 1]
```diff
@@ -15,13 +15,35 @@ import {
 } from 'graphql';
 
 export const Results = {
```
Comment (broskoski) [inline schema/home/home_page_artist_module.js]: Into this approach, very nice.

File: schema/home/home_page_artist_modules.js [hunk 2]
```diff
@@ -13,7 +14,15 @@ import {
 const HomePageArtistModules = {
   type: new GraphQLList(HomePageArtistModuleType),
   description: 'Artist modules to show on the home screen',
-  resolve: () => map(keys(Results), key => ({ key })),
+  resolve: (root, params, { rootValue: { accessToken, userID } }) => {
+    // First check each type if they can display…
+    return Promise.all(map(Results, ({ display }, key) => {
+      return display(accessToken, userID).then(displayable => ({ key, displayable }));
```
Comment (broskoski) [inline schema/home/home_page_artist_modules.js]: 🔥

File: test/schema/home/home_page_artist_module.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    it('returns suggestions', () => {
+      return runAuthenticatedQuery(query('suggested')).then(({ home_page }) => {
+        home_page.artist_module.results.should.eql([{ id: 'suggested' }]);
+      });
+    });
+  });
+
+  describe('when signed-out', () => {
+    shared();
```
Comment (broskoski) [inline test/schema/home/home_page_artist_module.js]: Hm what's the point of running this twice?

File: test/helper.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+  });
+}
+
+/**
+ * Same as `runQuery` except it provides a `rootValue` that’s required for authenticated queries.
+ *
+ * @see runQuery
+ */
+export function runAuthenticatedQuery(query) {
```
Comment (broskoski) [inline test/helper.js]: Nice helpers, thank you!

File: schema/home/home_page_artist_module.js [hunk 5]
```diff
@@ -15,13 +15,35 @@ import {
...
+          return results.length > 0;
+        });
+      });
+    },
+    resolve: (accessToken, userID) => {
+      if (accessToken && userID) {
+        return Results.suggested.fetch(accessToken, userID).then(results => {
+          return map(results, 'artist');
+        });
+      } else {
```
Comment (broskoski) [inline schema/home/home_page_artist_module.js]: You can just remove this `else` and `return Promise.resolve(null)` below.

**Comment by broskoski** [comment]
This looks awesome, I might steal your approach and refactor some of the homepage artwork modules with this nicer Promise situation. 

Looks like you have some ❌ 's from the linter, otherwise feel free to merge when ready.

File: test/schema/home/home_page_artist_module.js (hunk 3 above)
Comment (alloy) [inline test/schema/home/home_page_artist_module.js]: Good catch, they were supposed to run with and without authentication. Fixed that.

File: test/helper.js (hunk 4 above)
Comment (alloy) [inline test/helper.js]: :bow:

File: schema/home/home_page_artist_module.js (hunk 5 above)
Comment (alloy) [inline schema/home/home_page_artist_module.js]: Oooohhhh, nice. Done.

**Comment by alloy** [comment]
Ace, will merge then 👌

**Pull request merged**